### PR TITLE
Fix unique constraint and primary key generator issues for Postgres and H2

### LINKED ISSUE #1384: addPrimaryKey / addUniqueConstraint could support forIndexName on PostgreSQL
## Environment
**Liquibase Version**: 3.10.2

**Liquibase Integration & Version**: all

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: postgresql >= 9.1

**Operating System Type & Version**: all

## Description
Since version 9.1, PostgreSQL supports the same USING INDEX syntax when creating a unique constraint as Oracle does.

Therefore, the forIndexName could be supported by liquibase for addPrimaryKey and addUniqueConstraint

## Steps To Reproduce
```
        <addPrimaryKey columnNames="some_column" constraintName="some_constraint" tableName="some_table" forIndexName="some_index" />

```

## Actual Behavior
Changeset is rejected on postgresql.

## Expected/Desired Behavior
Changeset should be executed.

## Additional Context
It's probably enough to add postgresql to the checks at https://github.com/liquibase/liquibase/blob/31a22561423919b3875e0563a7bdcde3b9e457a9/liquibase-core/src/main/java/liquibase/sqlgenerator/core/AddPrimaryKeyGenerator.java#L35 and https://github.com/liquibase/liquibase/blob/d6dcd487fbb56a4a381491802c341f50e74bee59/liquibase-core/src/main/java/liquibase/sqlgenerator/core/AddUniqueConstraintGenerator.java#L33

### LINKED ISSUE #5410: SQL Unique constraint generation for h2 should not contain "USING INDEX" clause
### Search first

- [X] I searched and no similar issues were found

### Description

When generating a diff changelog with liquibase, the output generation of unique constraints contains "USING INDEX" for a h2 database output sql file when the reference database was a oracle database. This could be improved by adding a condition !h2Database to the generation code.

The generated code is:
```
CREATE UNIQUE INDEX FALL_FKRQ_IX ON FALL(FALL_KEY, QUARTAL);
ALTER TABLE FALL ADD CONSTRAINT UC_FALL_KEY_QUARTAL UNIQUE (FALL_KEY, QUARTAL) USING INDEX FALL_FKRQ_IX;
```

But it should be
```
ALTER TABLE FALL ADD CONSTRAINT UC_FALL_KEY_QUARTAL UNIQUE (FALL_KEY, QUARTAL);
```

It works fine, when using a h2 database as a reference database.

The usecase we have is to automatically generate incremental diff changeset for various databases (using the liquibase gradle plugin) based on our Spring / JPA Entities compared to a reference Database (oracle).

### Steps To Reproduce

- Create a JPA entity with a @UniqueConstraint annotation
- Use a spring url as a reference url
- Use a oracle jdbc url as a url
- Set the changelog filename to anything.h2.sql
- Generate diff changelog

### Expected/Desired Behavior

Liquibase omits the "USING INDEX" clause when the database detected based on the filename is h2.

### Liquibase Version

4.24.0

### Database Vendor & Version

Oracle, H2

### Liquibase Integration

gradle

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

Windows 10

### Additional Context

The fix would be to change
```
if (statement.getForIndexName() != null) {
            sql += " USING INDEX ";
            sql += database.escapeObjectName(statement.getForIndexCatalogName(), statement.getForIndexSchemaName(),
                statement.getForIndexName(), Index.class);
            isInUsingIndexClause = true;
        }
```
to
```
if (statement.getForIndexName() != null && !(database instanceof H2Database)) {
            sql += " USING INDEX ";
            sql += database.escapeObjectName(statement.getForIndexCatalogName(), statement.getForIndexSchemaName(),
                statement.getForIndexName(), Index.class);
            isInUsingIndexClause = true;
        }
```
in liquibase.sqlgenerator.core.AddUniqueConstraintGenerator#generateSql

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by Guschtel**
…the "using index" clause for the h2 database

## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 

## Description

Fixes #5410 
Fixes #1384

### CODE REVIEW

**Comment by Guschtel** [comment]
Fixes #5410

@MalloD12 can you please advise if that is what you are expecting?

**Comment by MalloD12** [comment]
Hi @Guschtel,

Code changes look good to me.  The only thing I would ask is whether it's possible to instead of creating a Java class for tests we do it as a groovy test class. For new test suites, we tried to use that language.

Test would look like:

```
 @Unroll
 def  "test unique constraint using index for #database database"() {
 given:
        AddUniqueConstraintStatement statement = new AddUniqueConstraintStatement(null, null, TABLE_NAME, new ColumnConfig[]{new ColumnConfig().setName(COLUMN_NAME), new ColumnConfig().setName(COLUMN_NAME2)}, CONSTRAINT_NAME);
        statement.setForIndexName(INDEX_NAME);
then:
            assertEquals(expectedSql, this.generatorUnderTest.generateSql(statement, database, null)[0].toSql(), "Failed for " + database.getShortName()))
where:
database | expectedSql
new H2Database() |  "ALTER TABLE AddUQTest ADD CONSTRAINT UQ_TEST UNIQUE (colToMakeUQ, colToMakeUQ2)"
....       
    }
```

Thanks,
Daniel.

**Comment by Guschtel** [comment]
@MalloD12 have a look again, I tried to change the tests to Spock tests. First time using Spock, so please be kind if there's room to improve. :D
I didn't really find a way to parameterize the tests as i'm used to do it with JUnit, so i flattened out all possible cases.

I Also added the changes from #1384 and testcases for that.

**Review by MalloD12 — APPROVED** [review]
**Approved**.

Code changes look good to me. Thank you @Guschtel for submitting this PR to fix two issues. Also, thank you for changing the tests format and use Spock (Groovy).

Build and most of the tests have been successfully executed (except functional test which seems to be failing for a non-related reason).

Thanks,
Daniel.

File: .mvn/jvm.config [hunk 1]
```diff
@@ -1,3 +0,0 @@
--Dhttp.keepAlive=false
```
Comment (filipelautert) [inline .mvn/jvm.config:1]: Should this maven wrapper file be removed as part of this pr?

File: liquibase-standard/src/main/java/liquibase/sqlgenerator/core/AddPrimaryKeyGenerator.java [hunk 2]
```diff
@@ -37,7 +37,7 @@ public ValidationErrors validate(AddPrimaryKeyStatement addPrimaryKeyStatement,
             }
         }
 
-        if (!((database instanceof OracleDatabase) || (database instanceof AbstractDb2Database))) {
+        if (!((database instanceof OracleDatabase) || (database instanceof AbstractDb2Database) || (database instanceof PostgresDatabase))) {
```
Comment (filipelautert) [inline liquibase-standard/src/main/java/liquibase/sqlgenerator/core/AddPrimaryKeyGenerator.java:40]: Please update PR title to reflect the changes to Postgresql too.

**Review by filipelautert — CHANGES_REQUESTED** [review]

File: liquibase-standard/src/main/java/liquibase/sqlgenerator/core/AddPrimaryKeyGenerator.java (hunk 2 above)
Comment (MalloD12) [reply liquibase-standard/src/main/java/liquibase/sqlgenerator/core/AddPrimaryKeyGenerator.java:40]: Done

File: .mvn/jvm.config (hunk 1 above)
Comment (MalloD12) [reply .mvn/jvm.config]: Done

**Review by filipelautert — APPROVED** [review]

**Review by rberezen — APPROVED** [review]

**Pull request merged**